fix: add the migration guide and nunjucks depreciation notes (#1253)
asyncapi/generator: apps/generator/docs/nunjucka-depreciate.md

--

> title: "Depreciation of nunjucks render engine"

Reviewer: ```suggestion
title: "Depreciation of Nunjucks render engine"
```

Reviewer 2: also spelling error in file name

Reviewer 2: title is a migration guide not deprecation notice

@Florence-Njeri shouldn't it sound completely different?

Reviewer: Hey @derberg , you're right the title is wrong

weight: 170
---

> # Migration Guide from nunjucks render engine to react render engine

Reviewer: ```suggestion
# Migration guide from Nunjucks render engine to React render engine
```

Reviewer 2: do we need h1 here, and in the other cli deprecation PR?

in website title is rendered as h1 header right? so with such manual h1 header it will look kinda duplicated, no?

## Introduction

> AsyncAPI Generator is moving away from Nunjucks templates in favor of React templates. This guide will help you migrate your existing Nunjucks templates to React.

Reviewer: ```suggestion
The AsyncAPI generator is moving away from Nunjucks templates in favor of React templates. This guide will help you migrate your existing Nunjucks templates to React.
```

> ## Step-by-Step Migration Guide

Reviewer: ```suggestion
## Step-by-step migration guide
```
Hey @Gmin2 We use sentence case for all the docs headings. Please re-write the rest of the headings in sentence case

### 1. Update package.json

> Change your template configuration in `package.json`:

Reviewer 2: and please explain that once deprecation period is ended, and we remove default nunjucks engine, react will become default and this setting will no longer be needed to configure

```json
{
"generator": {
"renderer": "react"
}
}
```

### 2. Install Dependencies

Install the necessary React dependencies:

```bash
npm install @asyncapi/generator-react-sdk
```

> ### 3. Basic Template Structure

Reviewer: ```suggestion
### 3. Basic template structure
```

Nunjucks:
```jsx
<h1>{{ asyncapi.info().title() }}</h1>
<p>{{ asyncapi.info().description() }}</p>
```

React:
```jsx
import { File } from '@asyncapi/generator-react-sdk';

export default function({ asyncapi }) {
  return (
    <File name="index.html">
      <h1>{asyncapi.info().title()}</h1>
      <p>{asyncapi.info().description()}</p>
    </File>
  );
}

> ```

Reviewer 2: also explain filenames

in nunjucks filename of template would have to be `index.html` and in react `index.html.js` or anything else with `*.js` pattern. Basically make it clear that in react filename of generated file is controlled by `File` component.

@Florence-Njeri please explain how to use website code component here to highlight specific part of the code - if you think it is needed.

> ### 4. Macros

Reviewer 2: Macros and Partials -> https://github.com/asyncapi/generator/blob/master/apps/generator/docs/nunjucks-render-engine.md?plain=1

Replace macros with React components:

Nunjucks:

> ```jsx

Reviewer 2: not jsx

{% macro renderChannel(channel) %}
  <div class="channel">
    <h3>{{ channel.address() }}</h3>
    <p>{{ channel.description() }}</p>
  </div>
{% endmacro %}

{{ renderChannel(someChannel) }}
```

React:
```jsx
// components/Channel.js
import { Text } from '@asyncapi/generator-react-sdk';

export function Channel({ channel }) {
  return (
    <Text>
      <div className="channel">
        <h3>{channel.address()}</h3>
        <p>{channel.description()}</p>
      </div>
    </Text>
  );
}

// Main template
import { File, Text } from '@asyncapi/generator-react-sdk';
import { Channel } from './components/Channel';

export default function({ asyncapi }) {
  return (
    <File name="channels.html">
      <Text>
        <h2>Channels</h2>
      </Text>
      {asyncapi.channels().map(channel => (

>         <Channel key={channel.address()} channel={channel} />

Reviewer 2: why `key` ?

      ))}
    </File>
  );
}
```

### 5. File template 

> //TODO: we can add a link to Florence docs once it is merged

Reviewer: Hey @Gmin2 You can already link the doc since I updated the current File template doc

Author: done !

> ## Testing Your Migration

Reviewer: ```suggestion
## Testing your migration
```

After migrating, test your template thoroughly:

> 1. Run the generator with your new React template

Reviewer: ```suggestion
1. Run the generator using your new React template
```

2. Compare the output with the previous Nunjucks template output
3. Check for any missing or incorrectly rendered content

## Conclusion

> Migrating from Nunjucks to React templates may require some initial effort, but it will result in more maintainable. You can read why we introduced react render engine [here](https://www.asyncapi.com/blog/react-as-generator-engine)

Reviewer: ```suggestion
Migrating from Nunjucks to React templates may require some initial effort, but it will result in more maintainable code. You can learn more about why we introduced the React render engine [here](https://www.asyncapi.com/blog/react-as-generator-engine).
```
